src/python/gem5/prebuilt/viper/cpu_cache_hierarchy.py

```diff
@@ -0,0 +1,273 @@
+
+            cores[i].connect_icache(cache.sequencer.in_ports)
+            cores[i].connect_dcache(cache.sequencer.in_ports)
+
+            cores[i].connect_walker_ports(
+                cache.sequencer.in_ports, cache.sequencer.in_ports
+            )
+
+            # Connect the interrupt ports
+            int_req_port = cache.sequencer.interrupt_out_port
+            int_resp_port = cache.sequencer.in_ports
+            cores[i].connect_interrupt(int_req_port, int_resp_port)
+
+            if i + 1 < num_cores:
+                cores[i + 1].connect_icache(cache.sequencer1.in_ports)
+                cores[i + 1].connect_dcache(cache.sequencer1.in_ports)
+
+                cores[i + 1].connect_walker_ports(
+                    cache.sequencer.in_ports, cache.sequencer1.in_ports
+                )
+
+                # Connect the interrupt ports
+                cores[i + 1].connect_interrupt(int_req_port, int_resp_port)
+
+            self._controllers.append(cache)
+
+        # Create the CPU directory controllers
+        self._directory_controllers = []
+
+        # Automatically determine the numa bit. This can be changed to
+        # increase the number of bytes to each memory channel before
+        # going to the next channels
+        dir_bits = int(math.log(len(board.get_mem_ports()), 2))
+        block_size_bits = int(math.log(board.get_cache_line_size()))
+
+        for addr_range, port in board.get_mem_ports():
+            dir = ViperCPUDirectory(
+                self.ruby_system.network,
+                board.get_cache_line_size(),
+                addr_range,
+                port,
+            )
+            dir.ruby_system = self.ruby_system
+            dir.version = len(self._directory_controllers)
+            self._directory_controllers.append(dir)
+
+            dir.L3CacheMemory = RubyCache(
+                size=self._l3_size,
+                assoc=self._l3_assoc,
+                replacement_policy=TreePLRURP(),
+                resourceStalls=False,
+                dataArrayBanks=16,
+                tagArrayBanks=16,
+                dataAccessLatency=20,
+                tagAccessLatency=15,
+            )
+
+        # Create the DMA Controllers, if required.
+        self._dma_controllers = []
+        if board.has_dma_ports():
+            dma_ports = board.get_dma_ports()
+            for i, port in enumerate(dma_ports):
+                ctrl = ViperCPUDMAController(
+                    self.ruby_system.network, board.get_cache_line_size()
+                )
+                ctrl.dma_sequencer = DMASequencer(version=i, in_ports=port)
+
+                ctrl.ruby_system = self.ruby_system
+                ctrl.dma_sequencer.ruby_system = self.ruby_system
+
+                self._dma_controllers.append(ctrl)
+
+        # Create DMA Controllers requires for any devices in the system.
+        device_dmas = []
+        if board.get_devices() is not None:
```

Comment:
I've changed the type passed to incorporate_cache in this file so that this function is available (not sure if you have to do this python).  In the future it would be nice if something like get_devices or get_pci_devices was part of an abstract board.

Reply:
Yes, we should probably have that interface. The problem is that devices behave very differently (at least, that's been my experience... maybe things are better now?). 

That said, I think that making a prebuilt cache hierarchy only compatible with a specific type of board is OK (and absolutely required with how complex gem5 is). In fact, I really like that this is explicitly only compatible with certain types of boards. This way we aren't advertising that any board can be used with any hierarchy.